src/Symfony/Component/Finder/Iterator/LazyIterator.php

```diff
@@ -0,0 +1,30 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Finder\Iterator;
+
+/**
+ * @author Jérémy Derussé <jeremy@derusse.com>
+ */
+class LazyIterator implements \IteratorAggregate
+{
+    private $iteratorFactory;
+
+    public function __construct(callable $iteratorFactory)
+    {
+        $this->iteratorFactory = $iteratorFactory;
+    }
+
+    public function getIterator()
+    {
+        yield from ($this->iteratorFactory)();
```

Comment:
actually, im trying to reason when we'd benefit from yielding the factorized iterator. Does it make sense?

`return ($this->iteratorFactory)();` seems as lazy

Reply:
ok. it does when doing `Finder::append(new LazyIt($c))`

https://3v4l.org/r7tjI

Reply:
This wouldn't fix the issue, as the instance of the factorized iterator will be passed to the caller and wont be destruct.

note: the original issue is because \AppendIterator keep a reference to the $iterators, and the `RecursiveDirectoryIterator` keep the file opened until destruct.

Reply:
@jderusse I think this would deserve a comment in the code explaining this need, so that it is not lost in the future.